cmake/GenerateScripts.cmake

```diff
@@ -0,0 +1,168 @@
+# Functions and macros to generate upgrade and downgrade scripts
+
+# concatenate_files(<output> <file> ...)
+#
+# Concatenate a list of files into <output>
+function(concatenate_files OUTPUT_FILE)
+  foreach(_file ${ARGN})
+    file(READ ${_file} _contents)
+    file(APPEND ${OUTPUT_FILE} "${_contents}")
+  endforeach()
+endfunction()
+
+# generate_script(...)
+#
+# Generate a script file and install it into the script directory.
+#
+# VERSION <version>
+#    Version to use for shared libraryx.
+#
+# SCRIPT <file>
+#    File name for script file.
+#
+# OUTPUT_DIRECTORY <dir>
+#    Directory for where script file should be written. Defaults to
+#    CMAKE_CURRENT_BINARY_DIR.
+#
+# FILES <file> ...
+#    List of files to include, in order, in the script file.
+function(generate_script)
+  set(options)
+  set(oneValueArgs VERSION SCRIPT OUTPUT_DIRECTORY)
+  set(multiValueArgs FILES)
+  cmake_parse_arguments(_generate "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  if(NOT _generate_OUTPUT_DIRECTORY)
+    set(_generate_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+  endif()
+
+  # Set the necessary variables, generate the real files, and append
+  # the output to the final update script. Ideally, the template
+  # files should have a '.in' suffix so that we can keep the
+  # template file and the final file available for debugging, but
+  # for some reason it was decided to not use the '.in' suffix for
+  # template files so now we're stuck with it.
+  set(MODULE_PATHNAME "$libdir/timescaledb-${_generate_VERSION}")
+  set(LOADER_PATHNAME "$libdir/timescaledb")
+
+  # Process all files. They are template files and should end with
+  # '.in' but for some reason they do not, so we append '.gen'
+  # instead to indicate generated files.
+  set(_result_files)
+  foreach(_file ${_generate_FILES})
+    configure_file(${_file} ${_file}.gen @ONLY)
```

Comment:
If the input file is in the source directory the generated file will also be in the source directory. An improvement could be to ensure that the generated file is in the build directory regardless of where the input file is.